### ISSUE TITLE
OC4 install extension

### ISSUE THREAD

**Issue by chulcha** [opencart/opencart#9106]
I think you need to add to the installation of the code field, which will carry the name of the module to make it requried.

```
<?xml version="1.0" encoding="UTF-8"?>
<modification>
    <name>Module name</name>
    <version>module version</version>
    <author>owner of module</author>
    <code>namespace</code>
<!--
extension\[namespace]\...
-->
    <link></link>
</modification>
```

Now the namespace is determined by the file name, which in my opinion is not correct, and many conflicts can arise.

**Comment by Hkr32** [comment]
I would do a slightly different structure:
I would also add a folder with the name of the module after the author's folder.
If I need to remove an extension, I went to the extensions folder, then to the author's folder and deleted the module folder, and not go through the heap of folders to delete each file by hand.
For example:
* extension/opencart/analytics/{admin, catalog}/*
* extension/hkr/my_extension_name/{admin, catalog}/*

**Comment by osworx** [comment]
In general we should think about the "sense" of the install.xml file and its content.
Basically with OC 4.x no modifications are possible (because of events) - which is a very good decision.
But the main purpose of the install.xml is not given anymore - the modifications itself.

Currently any extension can be installed providing only the "upload" folder - no install.xml is needed.
So, what for do we need this install.xml furthermore?
Maybe it helps developers and the store owner to see what is installed, because when no install.xml is given, no entries in the extension list.
Events have to be checked additionally to see what has been installed beside the core.

The argument to check the extensons folder(s) and so on is useless, because since 3.x the system stores the paths of folders and files and remove them automatically at uninstall.
But (as other systems doing), new tags about the structure (folder, subfolders and single files) could be added to the install.xml.
Doing so and storing the folders and files in the db is not needed anymore - just more work for developers (well most of them earn money, so this extra work should eb easy to fullfill).

And beside of possible new tags, I would add also a new tad <description> .. </description> which should contain the main purpose of the extension.
Because some of the extensions (and especially when a store has tons of them installed) have titles, which are useless and do not help me finding what it does!

Daniel should rework in general this install.xml and the content of it.
To have finally a usefull file (and content).
Beside this, the structure of it should be clearly communicated and proven at the installation process.

The modification list should be the "one stop place" of all additionally installed extension and therefore a help for the users (and devs).

Just my thoughts ..

**Comment by SaifAlSubhi** [comment]
@osworx question, please
"Currently any extension can be installed providing only the "upload" folder"
Do you mean that the modified php or twig files which are installed in /extension/ should be loaded over the core ones?

**Comment by osworx** [comment]
> 
> 
> "Currently any extension can be installed providing only the "upload" folder"
> Do you mean that the modified php or twig files which are installed in /extension/ should be loaded over the core ones?
>

What I am saying is, that an Extension **must** contain the upload folder and **can** contain the install.xml.
If an Extension has no install.xml, it will be not recorded in the list of the installed extensions (modifications).
The folders and files from the upload folder are copied directly (if they are inside the allowed folders which was introduced with 3.0.2.x).

Which means, any Extension without an install.xml and with upload folder only, is not visible (that it has been installed).
Except it has events - which means any dev (and store owner if he knows how to) have to controller both: the extension list AND the events for any modification.
But if an Extension does not has any event and no install.xml .. well I leave it up to your phantasy what it could do in the background ..

**Comment by chulcha** [comment]
> "Currently any extension can be installed providing only the "upload" folder"
No!!!! Need install.xml


The "sense" of the install.xml file and its content - is basic info about extention.
May be - description also.

May contain instructions for managing file uploads, such as scripts and styles of popular libraries

**Comment by osworx** [comment]
> 
> 
> > "Currently any extension can be installed providing only the "upload" folder"
> > No!!!! Need install.xml

Not correct, any Extension without an install.xml file can be installed.
Or how do you think I am installing some of my Extensions which need no modifcations (because of events)?

**Comment by chulcha** [comment]
@osworx 
https://github.com/opencart/opencart/blob/b3413606582619c654debdb930639218ef4e4b0a/upload/admin/controller/marketplace/installer.php#L165-L233

Where???

**Comment by osworx** [comment]
Where what?
How many Extensions shall i send you without any install.xml are installable?
The mentioned code part shows only that IF a xml-file is present to populate those infos directly into the database - nothing else and more.

But you can test that simply by yourself.
Just create a simple extension with upload folder only.
Name it my-useless-extension.ocmod.zip
And install ..

**Comment by chulcha** [comment]
I'm talking about OC4, and what are you talking about?

**Comment by osworx** [comment]
I am talking about the installer and that an Extension does NOT need a install.xml
And the whole discussion here IS about OC 4 (nevertheless, extension without install.xml can eb installed in OC 3.x).

**Comment by chulcha** [comment]
I focus on the 4th version
  and I support the require availability of an xml file

**Comment by osworx** [comment]
> 
> 
> I focus on the 4th version
> and I support the require availability of an xml file

Fine.
Then you you have to rewrite the code - curious what Daniel will say ..

**Comment by chulcha** [comment]
You don't need to rewrite anything.
It should only be added as mandatory.

Now the extension is being unpacked by filename
I suggest unpacking by code name

see 
$extension_data = [] sa data for method

**Comment by cgcmart** [comment]
I think xml file is needed, otherwise, admin panel extensions won't show uploaded modules because DB did not add them to related tables.
opencart.ocmod.zip contains install.xml.
Somewhere Daniel says, opencart default extensions do not need xml file because they are added to sql, let Opencart fresh installation install them.

**Comment by osworx** [comment]
As already said: the install.xml is currently (and in OC 4.x) not needed.
I am installing each day several of my own extension without - so that works perfectly.
Maybe not in the sense of Daniel, but - also as written above, the original sense of the install.xml was to add the modifications .. which are not possible anymore (due OC 4.x).

So I do not think, I know that.
The part you are correct is, that no entries are made in the DB.

Basically it is the same with the install (php & sql) files OC 2.x had in the past - they are also obsolete.

**Comment by osworx** [comment]
> Now the extension is being unpacked by filename
> I suggest unpacking by code name

No, as can be seen currently, some "developers" use somewhat.
And some so called "developers" should not publish any Extension!
Therefore I would suggest a "main repositiory" where all Extensions are registered.
And each Extension will get a unique number under which it will be populated to the database.
If not registered, no installation.

With such a system Opencart will have the control and could prevent users from installing useless extensions (or whatever they are called then).

**Comment by chulcha** [comment]
And who are the judges? (с) Griboyedov 

Who will moderate the usefulness / uselessness of the module
See how many duplicates, at least free ones

**Comment by Hkr32** [comment]
instal.xml needs to be reworked and made more informative and mandatory for every extension.

**Comment by osworx** [comment]
Who needs "judges"?
We need only correct coded Extensions, following the guidelines and GNU/GPL principles!
No crap needing ioncube and other encrypted code!

And duplicates can happen, why not?
While I mean with "duplicate" the same Idea, NOT the same and stolen Extension.

The system can be a "automated" one, with quality checks and so on.
Currently the Extension Market has too many of those lousy extensions providing only crap in the sense of encrypted fieles and/or advertising other services only.
Of course a "manual" check for NOT trusted Developers should be made - especially if a new developer will publish a new Extension.

We all (the developers) should not forget who is paying the bills at end - the customers/user who  are buying our extensions.
_So if you do not have to hide anything, you should vote for such a system._

**Comment by Hkr32** [comment]
> We need only correct coded Extensions, following the guidelines and GNU/GPL principles!

If I understand the GPL correctly, anyone can take your paid extension, modify it and sell it as their own, leaving your details to be credited to you as the author. Right?
https://www.gnu.org/licenses/gpl-faq.html#DoesTheGPLRequireAvailabilityToPublic

Therefore, many developers encrypt their code. Although I am personally against encryption. The official marketplace probably also has such extensions)))

**Comment by osworx** [comment]
> If I understand the GPL correctly, anyone can take your paid extension, modify it and sell it as their own, leaving your details to be credited to you as the author. Right?
> https://www.gnu.org/licenses/gpl-faq.html#DoesTheGPLRequireAvailabilityToPublic
Basically correct - that's the way it works.
But the misunderstanding IS, you can't sell others work as yours - you are allowed only to sell  the service around (e.g. hosting and providing the copy).
 
> Therefore, many developers encrypt their code. Although I am personally against encryption. The official marketplace probably also has such extensions)))
And that IS exactly the point those extensions are not conform with the GNU/GPL (and therefore not allowed).
You know - OpenSource .. ?!
And again, this IS also one reason why Extension in the official Marketplace MUST be controlled.

**Comment by Hkr32** [comment]
> But the misunderstanding IS, you can't sell others work as yours - you are allowed only to sell the service around (e.g. hosting and providing the copy).

Where is it said? I do not do this, I have a forum and several extensions, but when I studied the question, I realized that the point of the GPL is in freedom, that anyone can take someone else's extension, its source code, rework and sell it as his own revision. Therefore, I have not yet made paid extensions(((

**Comment by osworx** [comment]
**The Freedom in use (usage) and readable (= Open Source) code!**
_Not selling others work as yours - not publishing encrypted code._

If you ask and get the consent of the author - then yes.
You can also buy the Extension from the developer - then you have all therights he gives you (Copyright & / || Intellect).

But never sell others work as yours (parts or whole) - in such a case you will hear from the FSF.
And correct, another task, another discussion.

**Comment by condor2** [comment]
Right now you can't install extensions just by uploading because you can't see/access in admin. You need to add extra info in database manually in **extension_path** table.

**Comment by SaifAlSubhi** [comment]
Gentlemen, can someone see if this [#9112](https://github.com/opencart/opencart/pull/9112) should be right?

**Comment by Hkr32** [comment]
Maybe we should already use `json` instead of `xml`

**Comment by awaro** [comment]
> 
> 
> Who needs "judges"?
> We need only correct coded Extensions, following the guidelines and GNU/GPL principles!
> No crap needing ioncube and other encrypted code!
> 
> And duplicates can happen, why not?
> While I mean with "duplicate" the same Idea, NOT the same and stolen Extension.
> 
> The system can be a "automated" one, with quality checks and so on.
> Currently the Extension Market has too many of those lousy extensions providing only crap in the sense of encrypted fieles and/or advertising other services only.
> Of course a "manual" check for NOT trusted Developers should be made - especially if a new developer will publish a new Extension.
> 
> We all (the developers) should not forget who is paying the bills at end - the customers/user who are buying our extensions.
> _So if you do not have to hide anything, you should vote for such a system._

Varese will already exist, no matter what you do there, no matter what systems you do. they don't care.
It is better to bring opencart itself to mind first, than to engage in nonsense with all sorts of meaningless things.

**Comment by awaro** [comment]
Output breadcrumbs as a separate file and controller!
and don't repeat the old mistakes - go ahead!

**Comment by danielkerr** [comment]
to late in the game to start messing with this. not sure who the new members are recently but you seem to be reporting a lot of non bugs and feature requests very close to launching the next version. 

@Hkr32
no you can not sell or steal other peoples code! the extension that people are selling just uses a compatible opencart api. are you saying all android apps are free because the use they use the android api's to run.

**Comment by Hkr32** [comment]
@danielkerr, I am not selling other people's modules and am not going to do it, I just read the license and these conclusions from it. If you have a lawyer, you can address these questions to him.

**Comment by chulcha** [comment]
> to late in the game to start messing with this. not sure who the new members are recently but you seem to be reporting a lot of non bugs 

It has been written about repeatedly
How much effort was spent to make the pagination controller appear?
